native/src/routePlannerFrontEnd.cpp

```diff
@@ -227,6 +227,448 @@ vector<SHARED_PTR<RouteSegmentResult>> RoutePlannerFrontEnd::searchRouteInternal
+					if (routeFound) {
+						// route is found - cut the end of the route and move to next iteration
+						// start.stepBackRoute = new ArrayList<RouteSegmentResult>();
+						// boolean stepBack = true;
+						bool stepBack = stepBackAndFindPrevPointInRoute(gctx, gpxPoints, start, next);
+						if (!stepBack) {
+							// not supported case (workaround increase MAXIMUM_STEP_APPROXIMATION)
+							OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Info,
+											  "Consider to increase MAXIMUM_STEP_APPROXIMATION to: %f", routeDist * 2);
+							start->routeToTarget.clear();
+							routeFound = false;
+							break;
+						}
+					}
+				}
+				if (!routeFound) {
+					// route is not found move next point closer to start point (distance / 2)
+					routeDist = routeDist / 2;
+					if (routeDist < gctx->MINIMUM_STEP_APPROXIMATION &&
+						routeDist > gctx->MINIMUM_STEP_APPROXIMATION / 2 + 1) {
+						routeDist = gctx->MINIMUM_STEP_APPROXIMATION;
+					}
+					next = findNextGpxPointWithin(gpxPoints, start, routeDist);
+					if (next) {
+						routeDist = min(next->cumDist - start->cumDist, routeDist);
+					}
+				}
+			}
+		}
+		// route is not found skip segment and keep it as straight line on display
+		if (!routeFound) {
+			// route is not found, move start point by
+			next = findNextGpxPointWithin(gpxPoints, start, gctx->MINIMUM_STEP_APPROXIMATION);
+			if (prev) {
+				prev->routeToTarget.insert(prev->routeToTarget.end(), prev->stepBackRoute.begin(),
+										   prev->stepBackRoute.end());
+				makeSegmentPointPrecise(prev->routeToTarget[prev->routeToTarget.size() - 1], start->lat, start->lon,
+										false);
+				if (next) {
+					OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Info, "NOT found route from: %s at %d",
+									  start->pnt->getRoad()->getName().c_str(), start->pnt->getSegmentStart());
+				}
+			}
+			prev = nullptr;
+		} else {
+			prev = start;
+		}
+		start = next;
+	}
+	if (gctx->ctx->progress != nullptr) {
+		gctx->ctx->progress->timeToCalculate.Pause();
+	}
+	// BinaryRoutePlanner.printDebugMemoryInformation(gctx.ctx);
+	calculateGpxRoute(gctx, gpxPoints);
+	if (!gctx->result.empty() && !gctx->ctx->progress->isCancelled()) {
+		// new RouteResultPreparation().printResults(gctx.ctx, gpxPoints[0]->loc, gpxPoints[gpxPoints.size() - 1]->loc,
+		// gctx.result); System.out.println(gctx);
+	}
+	//+ if (resultMatcher != NULL) {
+	//+ resultMatcher.publish(gctx.ctx->progress->isCancelled() ? NULL : gctx);
+	//+ }
+	return gctx;
+}
+
+void makeSegmentPointPrecise(SHARED_PTR<RouteSegmentResult> routeSegmentResult, double lat, double lon, bool st) {
+	int px = get31TileNumberX(lon);
+	int py = get31TileNumberY(lat);
+	int pind = st ? routeSegmentResult->getStartPointIndex() : routeSegmentResult->getEndPointIndex();
+
+	SHARED_PTR<RouteDataObject> r = std::make_shared<RouteDataObject>(routeSegmentResult->object);
+	routeSegmentResult->object = r;
+	std::pair<int, int>* before = NULL;
```

Comment:
refrence is not needed

Reply:
The code could be changed to avoid NULL and weird pointers, better to use constant values (like negative int)